44385 Improve documentation on customizing the config screen (#345)
replicatedhq/replicated-docs: docs/vendor/config-screen-map-inputs.md

# Mapping User-Supplied Values

This topic describes how to map the values that your users provide in the Replicated admin console configuration screen to your application.

This topic assumes that you have already added custom fields to the admin console configuration screen by editing the Config custom resource. For more information, see [Creating and Editing Configuration Fields](admin-console-customize-config-screen).

## Overview of Mapping Values

> You use the values that your users provide in the admin console configuration screen to render YAML in the custom resource manifest files for your application.

Reviewer: small clarification - not all user-supplied yaml manifests are Custom Resources

```suggestion
You use the values that your users provide in the admin console configuration screen to render YAML in the manifest files for your application.
```

For example, if you provide an embedded database with your application, you might add a field on the admin console configuration screen where users input a password for the embedded database. You can then map the password that your user supplies in this field to the Secret manifest file for the database in your application.

Similarly, you can include fields on the configuration screen where your users can enable a custom ingress controller for the cluster. You then map these user-supplied values to the Ingress custom resources in your application.

For a tutorial of mapping database configuration options in a sample application, see [Adding Database Configuration Options for your Application](tutorial-adding-db-config).

> For an example of adding custom Ingress resources based on user-supplied configuration, see [Configuring Cluster Ingress](packaging-ingress).

Reviewer: Nice! This is a great guide that doesn't get enough looks. Glad we're linking to it.

1. In the [vendor portal](https://vendor.replicated.com/apps), click **Releases**. Then, click **View YAML** next to the desired release.

1. Open the Config custom resource manifest file that you created in the [Add Fields to the Configuration Screen](admin-console-customize-config-screen#add-fields-to-the-configuration-screen) procedure. The Config custom resource manifest file has `kind: Config`.

1. In the Config manifest file, locate the name of the user-input field that you want to map.

   **Example**:
   ```yaml
   spec:
    groups:
      - name: example_group_name
        title: "Example Group Title"
        description: "An example group of configuration options.""
        items:
          - name: example_config_screen_field_name
            title: "Example Field Title"
            type: bool
            default: "0"
   ```

   In the example above, the field name to map is `example_config_screen_field_name`.

1. In the same release in the vendor portal, open the custom resource manifest file where you want to map the value for the field that you selected.

   Use the ConfigOption template function to map the user-supplied value in a key value pair:

   ```yaml

>    example_key: '{{repl ConfigOption "CONFIG_SCREEN_FIELD_NAME"}}'

Reviewer: Why not use the field name from the example? `example_config_screen_field_name`

   ```
   Replace `CONFIG_SCREEN_FIELD_NAME` with the name of the field from the Config manifest file.

   **Example**:

   The following example shows mapping user-supplied TLS certificate and TLS private key files to the `tls.cert` and `tls.key` keys in a Secret custom resource manifest file.

   ```yaml
   apiVersion: v1
   kind: Secret
   metadata:
     name: tls-secret
   type: kubernetes.io/tls
   data:
     tls.crt: '{{repl ConfigOption "tls_certificate_file" }}'
     tls.key: '{{repl ConfigOption "tls_private_key_file" }}'
   ```

> 

Reviewer: This is another great example. I might add a link back to the ingress guide because the last option where we re-use the `kotsadm-tls` secret should be considered before a user goes in marking up their own custom TLS ConfigOption Fields.

```suggestion
 **Note**: While the example above is perfectly valid, if you're specifically looking to handle TLS secrets, check out the last example in [Configuring Cluster Ingress](packaging-ingress). Depending on your use case, there are some shortcuts you can take to re-use the certificates uploaded for the App Manager console itself.
```

   For more information about the ConfigOption template function, see [Config Context](../reference/template-functions-config-context#configoption) in the _Template Functions_ section.

1. Save and promote the release to a development environment to test your changes.

### Map Values to a Helm Chart

> The `values.yaml` file in a Helm chart contains values that are specific to the end-user environment. With Replicated, your users provide these values through the configuration screen in the admin console. You customize the configuration screen based on the required and optional configuration fields that you want to expose to your users.

Reviewer: ```suggestion
The `values.yaml` file in a Helm chart defines parameters that are specific to the end-user environment. With Replicated, your users provide these values through the configuration screen in the admin console. You customize the configuration screen based on the required and optional configuration fields that you want to expose to your users.
```

Reviewer: i dunno how you wanna wordsmith these but a few tiny tweaks

Author: Very nice

To map the values that your users provide in the admin console configuration screen to your Helm chart `values.yaml` file, you create a Replicated HelmChart custom resource.

To follow a tutorial that maps values from the configuration screen to a Helm chart, see [Example: Mapping the Configuration Screen to Helm Values](helm-mapping-example).

To map user inputs from the configuration screen to the `values.yaml` file:

> 1. In the [vendor portal](https://vendor.replicated.com/apps), click **Releases**. Then, click **View YAML** next to the desired release.

Reviewer: @cremerfc @tdensmore @jdewinne do you folks think it will be obvious to advanced users that they can do this in their GH repo or local CLI env as well?

Author: You know, I was thinking about that as I was writing these... there are probably a few good ways to address this in the docs (that is, that vendors can accomplish tasks through a few different mediums, including the vendor portal, the cli, their gh repo). 

I do think we should be documenting these different paths. Curious to hear what you all think. If you agree, this could be another effort to bring these different paths in across the procedures in the docs.

Reviewer: I wonder if there's a generic way we can say "edit your manifests" without specifying which of the three levels of advanced-ness?

Author: Yeah, totally--one idea would be that we can have a single topic that describes how to do these three methods. Then, we can link to that topic from procedures like this, when there are few options for how to access/edit/release the manifests
